pairtools/pairtools_dedup.py

```diff
@@ -1,248 +1,421 @@
+)
 @click.option(
-    "--p1", 
-    type=int, 
-    default=_pairsam_format.COL_P1,  
-    help='Position 1 column; default {}'.format(_pairsam_format.COL_P1))
+    "--p1",
+    type=int,
+    default=_pairsam_format.COL_P1,
+    help="Position 1 column; default {}".format(_pairsam_format.COL_P1),
+)
 @click.option(
-    "--p2", 
-    type=int, 
-    default=_pairsam_format.COL_P2,  
-    help='Position 2 column; default {}'.format(_pairsam_format.COL_P2))
+    "--p2",
+    type=int,
+    default=_pairsam_format.COL_P2,
+    help="Position 2 column; default {}".format(_pairsam_format.COL_P2),
+)
 @click.option(
-    "--s1", 
-    type=int, 
-    default=_pairsam_format.COL_S1,  
-    help='Strand 1 column; default {}'.format(_pairsam_format.COL_S1))
+    "--s1",
+    type=int,
+    default=_pairsam_format.COL_S1,
+    help="Strand 1 column; default {}".format(_pairsam_format.COL_S1),
+)
 @click.option(
-    "--s2", 
-    type=int, 
-    default=_pairsam_format.COL_S2,  
-    help='Strand 2 column; default {}'.format(_pairsam_format.COL_S2))
+    "--s2",
+    type=int,
+    default=_pairsam_format.COL_S2,
+    help="Strand 2 column; default {}".format(_pairsam_format.COL_S2),
+)
 @click.option(
-    "--unmapped-chrom", 
-    type=str, 
-    default=_pairsam_format.UNMAPPED_CHROM,  
-    help='Placeholder for a chromosome on an unmapped side; default {}'.format(_pairsam_format.UNMAPPED_CHROM))
+    "--unmapped-chrom",
+    type=str,
+    default=_pairsam_format.UNMAPPED_CHROM,
+    help="Placeholder for a chromosome on an unmapped side; default {}".format(
+        _pairsam_format.UNMAPPED_CHROM
+    ),
+)
 @click.option(
-    "--mark-dups", 
+    "--mark-dups",
     is_flag=True,
     help='If specified, duplicate pairs are marked as DD in "pair_type" and '
-         'as a duplicate in the sam entries.')
+    "as a duplicate in the sam entries.",
+)
 @click.option(
-    "--extra-col-pair", 
+    "--extra-col-pair",
     nargs=2,
-    #type=click.Tuple([str, str]),
+    # type=click.Tuple([str, str]),
     multiple=True,
-    help='Extra columns that also must match for two pairs to be marked as '
-    'duplicates. Can be either provided as 0-based column indices or as column '
+    help="Extra columns that also must match for two pairs to be marked as "
+    "duplicates. Can be either provided as 0-based column indices or as column "
     'names (requires the "#columns" header field). The option can be provided '
-    'multiple times if multiple column pairs must match. '
-    'Example: --extra-col-pair "phase1" "phase2"'
-    )
-
+    "multiple times if multiple column pairs must match. "
+    'Example: --extra-col-pair "phase1" "phase2"',
+)
+@click.option(
+    "--save-parent-id",
+    is_flag=True,
+    help="If specified, duplicate pairs are marked with the readID of the retained"
+    "deduped read. Only has effect with scipy or sklearn backend",
+)
+@click.option(
+    "--backend",
+    type=click.Choice(["cython", "scipy", "sklearn"]),
```

Comment:
Does it make sense to import scipy at the top of the module if it's used only in case of "scipy" is selected? Can we probably move the imports for scipy to the line 467?

Reply:
Could do that, but it's the default and it's a hard dependency anyway... So maybe simpler to just import at the top?

Reply:
I actually made it a hard dependency in this branch, I just realised. But if we default to it, it should be a hard dependency, I guess...

Reply:
OK, makes sense!